### PR TITLE
Use categorical covariates properly and compute signature matrix in training only

### PR DESCRIPTION
**Pull request by SimonGrouard**
Two main features added in this PR : 
1. Compute categorical covariates properly. Before that, cell types were added in the categorical covariates, preventing to use real categorical covariates to be corrected for in the model. They are now added to the labels variable instead. More information can be found [here](https://docs.google.com/document/d/1OoGuzzlJGsUNueymKWBOUXoY8HBXQuauLFpfyTbzD1Y/edit#heading=h.3sn796mprzvn), Jan 24 2024.
2. When in evaluation, the latent signature matrix now used will be the last training latent signature computed (the one computed with the last training batch), so that we don't leak evaluation data to infer the proportions for the deconvolution metric, and over-estimate this metric.

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -27,15 +27,14 @@
...
 TRAIN_SIZE = 0.7 # as opposed to validation
 CHECK_VAL_EVERY_N_EPOCH = None
 if TRAIN_SIZE < 1:
     CHECK_VAL_EVERY_N_EPOCH = 1
 # MixUpVI model hyperparameters
 CONT_COV = None  # list of continuous covariates to include
-ENCODE_COVARIATES = False  # should be always False for now, we don't encode cat covar
-ENCODE_CONT_COVARIATES = False  # True or False, whether to include cont covar
+CAT_COV = ["donor_id", "assay"] # list of categorical covariates to include
+ENCODE_COVARIATES = True # whether to encode cont/cat covars (they are always decoded)
```
Comment (khalilouardini) [inline constants.py:37]: "they are always decoded" => Can you point to the part of the code where you've seen this please ?

File: scvi/dataloaders/_data_splitting.py [hunk 2]
```diff
@@ -231,11 +231,10 @@ def setup(
         self.train_idx = indices[n_val : (n_val + n_train)]
         self.test_idx = indices[(n_val + n_train) :]
 
-        cat_keys = self.adata_manager.registry["setup_args"]["categorical_covariate_keys"]
         cell_type_key = None
-        for elem in cat_keys:
-            if elem == "cell_type" or elem == "cell_types_grouped":
-                cell_type_key = elem
+        if (labels_key := self.adata_manager.registry["setup_args"]["labels_key"]) == \
+            "cell_type" or labels_key == "cell_types_grouped":
```
Comment (khalilouardini) [inline scvi/dataloaders/_data_splitting.py:236]: maybe we shoud use "cell_type" by default and rename it in the `adata` ?

File: tuning_configs.py [hunk 3]
```diff
@@ -76,13 +75,13 @@
...
     "pseudo_bulk": PSEUDO_BULK,
     "encode_covariates": ENCODE_COVARIATES,
-    "encode_cont_covariates": ENCODE_CONT_COVARIATES,
     "mixup_penalty": MIXUP_PENALTY,
     "dispersion": DISPERSION,
     "gene_likelihood": GENE_LIKELIHOOD,
     "train_size": TRAIN_SIZE,
     "batch_size": BATCH_SIZE,
     "n_latent": LATENT_SIZE,
+    "n_cell_types": len(GROUPS[TRAINING_CELL_TYPE_GROUP]) - 1,
```
Comment (khalilouardini) [inline tuning_configs.py:84]: what is the `GROUPS` variable?

**Review by khalilouardini — COMMENTED** [review]
Thanks Simon, a few minor comments!

File: constants.py (hunk 1 above)
Comment (SimonGrouard) [reply constants.py:37]: Yes! 
In the initialization of the VAE decoder [there](https://github.com/scverse/scvi-tools/blob/9b99f95be35120046c14ca69bd44f10c2d2188ee/scvi/module/_vae.py#L217) you can see that the categorical variables passed is cat_list. 
But in the initialization of the VAE encoder [there](https://github.com/scverse/scvi-tools/blob/9b99f95be35120046c14ca69bd44f10c2d2188ee/scvi/module/_vae.py#L179), what is given to the encoder is not exactly cat_list, but encoder_cat_list, which is equal to cat_list only when encoder_covariates is True.

File: scvi/dataloaders/_data_splitting.py (hunk 2 above)
Comment (SimonGrouard) [reply scvi/dataloaders/_data_splitting.py:236]: Yes indeed would be clearer, maybe in another PR ?

File: tuning_configs.py (hunk 3 above)
Comment (SimonGrouard) [reply tuning_configs.py:84]: The one where we define all the granularities, in constants.py.
So GROUPS[TRAINING_CELL_TYPE_GROUP] is the chosen granularity, and len(GROUPS[TRAINING_CELL_TYPE_GROUP]) - 1 is the number of cell types of this granularity.

**Review by khalilouardini — APPROVED** [review]

**Pull request merged**